Launch Crossgen2 via explicitly specified .NET host

trylek opened the pull request:
/cc @dotnet/crossgen-contrib 

Figuring out right now how to inject this into the runtime repo in order to actually test it, sending out in hope of early PR feedback.

Thanks

Tomas

trylek committed:
Launch Crossgen2 via explicitly specified .NET host

davidwrighton commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -208,6 +278,7 @@ private string GenerateCrossgen2ResponseFile()
             }
 
             result.AppendLine("-O");
+            result.AppendLine("--pdb");
```
We cannot safely pass --pdb to versions of crossgen2 before 6.0.

Some of the logic above abstracts the differences between 5.0 and 6.0 crossgen2 by checking the jitpath variable, and if it is set, but I think we should regularize this so there is a crossgen2 version that we are speaking of. That would allow us to be more intentional, such as specifying that if working with 5.0, when building pdbs, build them with crossgen.exe instead of crossgen2. When we get 6.0 we can use the `--pdb` switch. If we have a rigourous model where it is clear what code effects which versions of the compiler, I think it will be much easier to get this all correct.

trylek replied:
Based on your feedback I have modified the change so that --pdb and other "6.0 Crossgen2"-specific features are only triggered when JitPath is not specified. I have managed to validate this change locally after applying Steve's and Anirudh's fixes for rolling forward the runtime SDK drop to 6.0.

AntonLapounov replied:
Note that we do have Crossgen2 version number and it would be trivial to pass it through metadata: https://github.com/dotnet/sdk/blob/419b69015e49347ad9c5d1cc01e28bbd7c140394/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L143

davidwrighton commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -429,8 +429,8 @@ Copyright (c) .NET Foundation. All rights reserved.
-          Outputs="%(_ReadyToRunSymbolsCompileList.OutputPDBImage)">
-
+          Outputs="%(_ReadyToRunSymbolsCompileList.OutputPDBImage)"
+          Condition="'$(PublishReadyToRunUseCrossgen2)' != 'true'">
```
This isn't correct for crossgen2 on .NET 5.0. We should continue to create pdbs with crossgen when compiling targetting .NET 5.0 or earlier.

trylek replied:
Fixed in 3rd commit, thanks for pointing that out!

AntonLapounov commented:
Crossgen2 is published as a self-contained app. Why do we want to launch it using a host?

trylek committed:
Address David's PR feedback

trylek committed:
Fix condition for PDB generation

trylek commented:
I believe I have addressed the gist of David's feedback. For Anton's suggestion, as discussed offline this is non-trivial to do as publishing of the "self-contained Crossgen2 app" by itself needs the crossgenned framework as a prerequisite. Having said that, I don't think we can afford to wait until this fix propagates to the runtime repo by means of rolling forward the SDK version. The SDK upgrade is scheduled to be merged in as part of the March infra roll-out, please see

https://github.com/dotnet/runtime/pull/47245

so that one more upgrade (to include changes in this PR) will likely happen in April if not later and that's far too late for our other CG2 efforts. Apart from finalizing this change I plan to try to address or mitigate reviewer feedback on my other PR

https://github.com/dotnet/runtime/pull/47349

so that we can start testing CG2-compiled SDK ASAP. One way or another, please take another look at this change when you have a chance.

Thanks

Tomas

davidwrighton commented:
@trylek, My concerns have been addressed, but I'd rather somebody else walk through and verify that the behavior is correct as I'm generally not familiar with this codebase. I'd like to see @AntonLapounov provide a review.

AntonLapounov reviewed (changes requested):
I think all error messages should go to resource files to allow localization.

trylek committed:
Make error messages localizable per Anton's PR feedback

AntonLapounov commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -698,4 +698,52 @@ The following are names of parameters or literal values and should not be transl
     <comment>{StrBegin="NETSDK1148: "}</comment>
   </data>
+  <data name="CrossgenToolMissingInPDBCompilationMode" xml:space="preserve">
+    <value>NETSDK1149: CrossgenTool not specified in PDB compilation mode</value>
```
It seems the convention here is to end each error message with a period.

trylek replied:
Fixed in 6th commit, thanks for pointing that out.

AntonLapounov commented on src/Tasks/Common/Resources/xlf/Strings.de.xlf:
```diff
@@ -4,745 +4,805 @@
       <trans-unit id="AddResourceWithNonIntegerResource">
         <source>NETSDK1076: AddResource can only be used with integer resource types.</source>
-        <target state="translated">NETSDK1076: AddResource kann nur mit ganzzahligen Ressourcentypen verwendet werden.</target>
+        <target state="new">NETSDK1076: AddResource can only be used with integer resource types.</target>
```
Something went wrong in this file.

trylek replied:
I have no idea why it dropped all the German strings. I ended up reverting the file to its original content and just selectively injecting the new untranslated strings. I have no idea what happens the next time someone tries to add a resource string.

trylek replied:
OK, so after the manual restoration it managed to add new strings fine without dropping all the translation so it hopefully was just a one-off fluke local to my box.

trylek committed:
More PR feedback, resolve against current master

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -35,25 +35,34 @@ public class RunReadyToRunCompiler : ToolTask
 
         private bool IsPdbCompilation => !String.IsNullOrEmpty(_createPDBCommand);
 
+        private string DotNetHostPath => Crossgen2Tool?.GetMetadata("DotNetHostPath") ?? null;
```
Does the `?? null` part do anything?

trylek replied:
That's obviously no-op, removed in 6th commit, thanks for spotting!

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -65,33 +74,72 @@ protected override bool ValidateParameters()
             if (IsPdbCompilation && CrossgenTool == null)
             {
-                // We need the crossgen tool for now to emit native symbols. Crossgen2 does not yet support this feature
+                // PDB compilation is a step specific to Crossgen1 and 5.0 Crossgen2
```
What happens if we are using Crossgen2 version ≥ 6 and `IsPdbCompilation` is `true`?  That seems a possible combination unless we also change code in PrepareForReadyToRunCompilation.cs.

trylek replied:
I'll fix this in PrepareReadyToRunCompilation, Crossgen2 is currently not capable of lazy PDB / PerfMap generation (R2RDump currently supports that) and the scripts don't actually need that.

trylek replied:
Hopefully fixed in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -65,33 +74,72 @@ protected override bool ValidateParameters()
+                }
+                else
+                {
+                    if (string.IsNullOrEmpty(Crossgen2Tool.GetMetadata(MetadataKeys.TargetOS)))
+                    {
+                        Log.LogError(Strings.TargetOSMissingWhenJitPathNotSet);
+                        return false;
+                    }
+                    if (string.IsNullOrEmpty(Crossgen2Tool.GetMetadata(MetadataKeys.TargetArch)))
+                    {
+                        Log.LogError(Strings.TargetArchMissingWhenJitPathNotSet);
+                        return false;
+                    }
```
We could use a single parameterized error message and pass either `'TargetOS'` or `'TargetArch'`.

trylek replied:
Good point, fixed in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -103,12 +151,14 @@ protected override bool ValidateParameters()
                 // R2R image has to be created before emitting native symbols (crossgen needs this as an input argument)
                 if (String.IsNullOrEmpty(_outputPDBImage) || !File.Exists(_outputR2RImage))
                 {
+                    Log.LogError(Strings.PDBGeneratorInputExecutableNotFound, _outputR2RImage);
```
This error message would be confusing in case the R2R image is specified correctly, but `_outputPDBImage` is `null`.

trylek replied:
Good point, I have split this into two conditions with two error messages.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -149,16 +200,36 @@ private string GetAssemblyReferencesCommands()
 
+        protected override string GenerateCommandLineCommands()
+        {
+            if (UseCrossgen2 && !string.IsNullOrEmpty(DotNetHostPath))
```
Should we also check `&& !IsPdbCompilation`? BTW, it might be cleaner to have a flag like this:
```C#
private bool ActuallyUseCrossgen2 => UseCrossgen2 && !IsPdbCompilation;
```

trylek replied:
Added in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -149,16 +200,36 @@ private string GetAssemblyReferencesCommands()
+
+        private static string Quote(string path)
+        {
+            if (string.IsNullOrEmpty(path) || (path[0] == '\"' && path[path.Length - 1] == '\"'))
```
I do not think we should introduce this function.  We have already checked that `File.Exists(Crossgen2Tool.ItemSpec) == true`, which means `Crossgen2Tool.ItemSpec` cannot be `null`, an empty string, or quoted. Furthermore, quoting arbitrary paths (including directory paths) would require special handling of backslashes before quotes, which we'd better avoid.

trylek replied:
Removed in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -208,6 +279,20 @@ private string GenerateCrossgen2ResponseFile()
 
             result.AppendLine("-O");
+            
+            if (string.IsNullOrEmpty(jitPath))
```
As David mentioned, it would be cleaner to use Crossgen2 version check instead of introducing more occurrences of this dirty check.

trylek replied:
Hmm, while you pointed out how to access CG2 version in PrepareReadyToRunCompilation, I don't see any easy way to detect it in RunReadyToRunCompiler so we'd need to pass the version (or at least a boolean flag stating whether the version is 5 or not) as an extra argument and synthesize it somehow in the Microsoft.NET.Publish.targets script - that sounds like a bit of an overshoot to me.

AntonLapounov replied:
You can set it as a metadata in `ValidateCrossgen2Support`:
```C#
Crossgen2Tool.SetMetadata(MetadataKeys.Version, crossgen2PackVersion);
```
then use it in RunReadyToRunCompiler.cs and Microsoft.NET.Publish.targets instead of the `JitPath` check.

trylek replied:
Fixed in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -65,33 +74,72 @@ protected override bool ValidateParameters()
-                if (!File.Exists(CrossgenTool.ItemSpec) || !File.Exists(CrossgenTool.GetMetadata(MetadataKeys.JitPath)))
+                // We expect JitPath to be set for .NET 5 and {TargetOS, TargetArch} to be set for .NET 6 and later
+                string jitPath = Crossgen2Tool.GetMetadata(MetadataKeys.JitPath);
+                if (Crossgen2Tool == null || string.IsNullOrEmpty(Crossgen2Tool.ItemSpec))
```
Nit: In general we do not check `ItemSpec`s for `IsNullOrEmpty`, because it would require some special efforts to create and pass such an `ITaskItem`.  For instance, in this function we do not check that this field is not `IsNullOrEmpty`:
```C#
_outputR2RImage = CompilationEntry.ItemSpec
```

trylek replied:
Removed in 6th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -119,6 +169,7 @@ protected override bool ValidateParameters()
 
                 if (!File.Exists(_inputAssembly))
                 {
+                    Log.LogError(Strings.InputAssemblyNotFound, _inputAssembly);
```
Do we want to also check that `_outputR2RImage` is not `null` or an empty string?  Note that in this branch it is a metadata and not an `ItemSpec`.

trylek replied:
Good point, added in 6th commit.

trylek committed:
Address Anton's PR feedback

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -149,16 +220,25 @@ private string GetAssemblyReferencesCommands()
+        {
+            if (ActuallyUseCrossgen2 && !string.IsNullOrEmpty(DotNetHostPath))
+            {
+                return Crossgen2Tool.ItemSpec;
```
I didn't want to introduce the `Quote` function, but don't we still need to quote this one?
```C#
    return $"\"{Crossgen2Tool.ItemSpec}\"";
```

trylek replied:
I guess I misunderstood you, I thought you were saying that the quoting is undesirable as it's done internally in the base class. I'll fix that.

trylek replied:
Fixed in 7th commit.

trylek committed:
Quote Crossgen2Tool in GenerateCommandLineCommands

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -34,26 +35,38 @@ public class RunReadyToRunCompiler : ToolTask
+            get
+            {
+                string version5 = Crossgen2Tool?.GetMetadata(MetadataKeys.IsVersion5);
+                return version5 != null && bool.Parse(version5);
```
`bool.Parse` may throw, e.g., when you pass an empty string.
```suggestion
                return bool.TryParse(version5, out bool value) && value;
```

trylek replied:
Hmm, wouldn't it be safer to check IsNullOrEmpty and then make bool.Parse throw when garbage is passed in?

trylek replied:
Added empty string check in 8th commit.

AntonLapounov replied:
In project and targets file we usually use `'$(foo)' == 'true'` and `'$(foo)' != 'true'` comparisons and do not throw on incorrect Boolean values.  Note that in another place you use `TryParse` for the same check:
```C#
bool.TryParse(Crossgen2Tool.GetMetadata(MetadataKeys.IsVersion5), out crossgen2IsVersion5);
```
If you wish to have just the `!IsNullOrEmpty` check, I am fine with that.

trylek replied:
I have modified the other place in PrepareR2RC to use the same algorithm as Crossgen2IsVersion5 in RunR2RC. I do know that in msbuild scripts we typically check for "true" and I'm well acquainted with all the obscure bugs that can happen once someone accidentally passes an unexpected parameter value.

trylek committed:
Add empty string check to Crossgen2IsVersion5

trylek committed:
Put IsVersion5 parsing in PrepareR2RC in sync with RunR2RC

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -97,13 +97,51 @@ protected override void ExecuteCore()
+                    if (outputPDBImage != null && !crossgen2IsVersion5)
+                    {
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.EmitSymbols, "true");
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.OutputPDBImage, outputPDBImage);
```
It seems we set these metadata even for the old Crossgen.  Just wanted to make sure that was intentional.

trylek replied:
Not intentional, you're right, thanks for spotting, fixed in 9th commit.

trylek replied:
Well, 10th actually, time flies...

AntonLapounov reviewed (approved):
Thank you for cleaning up the code!

trylek committed:
Don't pass EmitSymbols / output PDB path to CG1

AntonLapounov commented on src/Tasks/Common/Resources/xlf/Strings.ko.xlf~RF26a08dc.TMP:
```diff
@@ -0,0 +1,815 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
I think this file should be removed.

trylek replied:
Damn it, I thought I had excluded it, thanks for spotting, hopefully should be gone now.

trylek committed:
Remove inadvertently added temporary file

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -97,13 +97,51 @@ protected override void ExecuteCore()
+                    {
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.EmitSymbols, "true");
+                        r2rCompilationEntry.SetMetadata(MetadataKeys.OutputPDBImage, outputPDBImage);
+                    }
```
We miss publishing the generated PDB file in Crossgen2 version 6 scenario.  Need to execute the code with `r2rSymbolsFileToPublish`.

trylek replied:
I see, thanks for the explanation, is it better now :-) ?

AntonLapounov replied:
That would not work for the composite mode, which generates a single PDB file.

trylek replied:
I have added a new section for publishing composite PDB / PerfMap, please take another look.

trylek committed:
Fix PDB publishing for CG2 6.0 per Anton's PR feedback

trylek committed:
Fix publishing PDB / PerfMap for composite R2R image

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -138,53 +209,31 @@ protected override void ExecuteCore()
-                        r2rFilesPublishList.Add(r2rSymbolsFileToPublish);
+                    // This TaskItem corresponds to the output PDB image. It is equivalent to the input TaskItem, only the ItemSpec for it points to the new path
+                    // for the newly created PDB image.
+                    TaskItem r2rSymbolsFileToPublish = new TaskItem(file);
+                    r2rSymbolsFileToPublish.ItemSpec = outputPDBImage;
```
Don't we want to skip this for the composite mode? We should publish just one PDB file.

trylek replied:
I have added a composite check here in 14th commit.

trylek committed:
One more fix for composite PDB publishing per Anton's suggestion

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -125,6 +168,34 @@ protected override void ExecuteCore()
+                    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                    {
+                        compositePDBImage = Path.ChangeExtension(compositeR2RImage, ".ni.map");
+                        compositePDBRelativePath = Path.ChangeExtension(compositePDBRelativePath, ".ni.map");
```
`compositePDBRelativePath` is `null` here. Shouldn't we use `compositeR2RImageRelativePath` instead?

trylek replied:
For sure, thanks for spotting!

trylek replied:
Fixed in 15th commit.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -97,13 +97,51 @@ protected override void ExecuteCore()
+                string outputPDBImageRelativePath = null;
+                string crossgen1CreatePDBCommand = null;
+
+                if (EmitSymbols)
```
What happens if `ReadyToRunUseCrossgen2`, `crossgen2IsVersion5`, `Crossgen2Composite`, and `EmitSymbols` are all true? Should we report an error?

trylek replied:
Good point, error reporting added in 15th commit.

trylek committed:
More Anton's PR feedback and other fixes

I have noticed that EmitSymbols is not handled consistently
- PrepareReadyToRunCompilation was setting it as metadata on the
compilation item whereas RunReadyToRunCompiler expected it as a
task parameter. I ended up converting RunReadyToRunCompiler to
use the compilation entry metadata as the logic to set the flag
in PrepareReadyToRunCompilation is non-trivial and trying to
duplicate it in the Microsoft.NET.Publish.targets script would be
ugly and error prone.

Thanks

Tomas

AntonLapounov commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -702,4 +702,60 @@ The following are names of parameters or literal values and should not be transl
+    <comment>{StrBegin="NETSDK1162: "}</comment>
+  </data>
+  <data name="Crossgen5CannotEmitSymbolsInCompositeMode" xml:space="preserve">
+    <value>NETSDK1163: Crossgen2 5.0 doesn't support emitting symbols in composite mode</value>
```
It misses a period at the end. Maybe rephrase it a little?
> Cannot emit symbols when publishing for .NET 5 with Crossgen2 using composite mode.

I assume if we ever have .NET 5.1, it will not support this scenario either.

trylek replied:
Good point, fixed in 16th commit. I have also fixed one other bug - CG2, just like CG1, expects just the directory names, not complete file names, in <code>--pdb-path</code> and <code>--perfmap-path</code> arguments, and that was handled incorrectly in PrepareForReadyToRunCompilation. This will ultimately need a CG2 fix to put perfmap naming in sync with what PrepareForR2RC expects, that is not the case today.

trylek committed:
Fix error message per Anton's feedback, fix PDB / PerfMap paths

The logic for emitting --pdb-path and --perfmap-path arguments
was incorrect as Crossgen2, just like Crossgen, only expects
the directory name, not the complete output file name for
PDB / PerfMap generation. I have already verified locally this
on Windows by injecting the fresh-built SDK into my runtime
clone with the CG2 publishing change, I'm now working on doing
the same on Linux.

Thanks

Tomas

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -208,6 +295,21 @@ private string GenerateCrossgen2ResponseFile()
+                else
+                {
+                    result.AppendLine("--perfmap");
+                    result.AppendLine("--perfmap-path:{_outputPDBImage}");
```
We miss `$` before the interpolated string here.

trylek committed:
Fix interpolated string per Anton's PR feedback

trylek commented:
@AntonLapounov - sigh, apparently I need to add some provisions to PrepareReadyToRunCompilation to cater for the fact that Crossgen2 until now hasn't been generating the MVID into the PerfMap name. I'll send out a PR for fixing that but that will only make it to Preview 2 - am I right to presume that we need something like a minor version check in the code?

AntonLapounov commented:
I don't think you need a minor version check. Make the fix in Runtime repo and wait a bit until it is picked up by the SDK repo. SDK will use the latest Crossgen2 package.

trylek commented:
Well, if I don't merge the SDK change in this week, it's going to slip to Preview 3 and its back-propagation to runtime will get delayed until Preview 3 ships (becomes generally available); I'm worried that pushes us uncomfortably close to the general .NET 6 shipping date.

AntonLapounov commented:
@trylek You may temporarily disable emitting perf maps in R2R SDK tests for .NET 6 to unblock this PR. Do I understand correctly that you want to place the MVID in the file name for non-composite mode only? Is having MVID in the file name important? Below are the errors I see in the log:
```
error MSB3030: Could not copy the file "obj/Debug/net6.0/ubuntu.16.04-x64/R2R/Crossgen2TestApp.ni.{4e16f2c1-0ed5-4bc0-88f3-2b978f1a98e5}.map" because it was not found.
error MSB3030: Could not copy the file "obj/Debug/net6.0/ubuntu.16.04-x64/R2R/ClassLib.ni.{233d699a-af57-4160-a632-0ad7769be9a8}.map" because it was not found.

error MSB3030: Could not copy the file "obj/Debug/net6.0/ubuntu.16.04-x64/R2R/Crossgen2TestApp.r2r.ni.map" because it was not found. 
```

trylek commented:
I guess that the purpose of the MVID in the perfmap name is similar to some checksum that's in the PDB to let diagnostic software verify that it matches, @hoyosjs or @brianrob might know. I guess you're right that I can disable the perfmap generation for now, it's just unfortunate to accrue debt.

AntonLapounov commented:
@trylek Is that a problem that for the composite perf map you are not using any MVID? If not, maybe omitting MVID in non-composite mode would be OK as well.

brianrob commented:
Right, the MVID is used to make sure that we only use a "matching" symbol file.  Sounds like for composite mode, we'd need to do something else, but perhaps we'd just depend on a recent enough version of the perf tools to be available on the box, and then it shouldn't matter, since we'd use the built-in Linux perf tool's jitdump functionality.

trylek committed:
Hotfix perfmap file name for .NET6 Preview1 version of Crossgen2

trylek committed:
Fix perfmap name for the composite case

trylek committed:
Add Patch check to properly set up PerfMap path for CG2 Preview1/2

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Nit: Did we remove the initial BOM?

trylek replied:
Sorry about that, hopefully restored now.

trylek committed:
Restore inadvertently removed UTF8 BOM marker

trylek committed:
Keep support for JitPath in .NET 6 CG2 to simplify the transition
